cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/adapter/AdapterService.java

```diff
@@ -0,0 +1,514 @@
+                           Map<String, String> defaultSourceProperties,
+                           Map<String, String> defaultSinkProperties,
+                           Map<String, String> defaultAdapterProperties,
+                           ProgramType programType) {
+      this.file = file;
+      this.type = adapterType;
+      this.sourceType = sourceType;
+      this.sinkType = sinkType;
+      this.defaultSourceProperties = ImmutableMap.copyOf(defaultSourceProperties);
+      this.defaultSinkProperties = ImmutableMap.copyOf(defaultSinkProperties);
+      this.defaultAdapterProperties = ImmutableMap.copyOf(defaultAdapterProperties);
+      this.programType = programType;
+    }
+
+    public File getFile() {
+      return file;
+    }
+
+    public String getType() {
+      return type;
+    }
+
+    public Source.Type getSourceType() {
+      return sourceType;
+    }
+
+    public Sink.Type getSinkType() {
+      return sinkType;
+    }
+
+    public Map<String, String> getDefaultSourceProperties() {
+      return defaultSourceProperties;
+    }
+
+    public Map<String, String> getDefaultSinkProperties() {
+      return defaultSinkProperties;
+    }
+
+    public Map<String, String> getDefaultAdapterProperties() {
+      return defaultAdapterProperties;
+    }
+
+    public ProgramType getProgramType() {
+      return programType;
+    }
+  }
+
+  private static class AdapterManifestAttributes {
+    private static final String ADAPTER_TYPE = "CDAP-Adapter-Type";
+    private static final String ADAPTER_PROPERTIES = "CDAP-Adapter-Properties";
+    private static final String ADAPTER_PROGRAM_TYPE = "CDAP-Adapter-Program-Type";
+    private static final String SOURCE_TYPE = "CDAP-Source-Type";
+    private static final String SINK_TYPE = "CDAP-Sink-Type";
+    private static final String SOURCE_PROPERTIES = "CDAP-Source-Properties";
+    private static final String SINK_PROPERTIES = "CDAP-Sink-Properties";
+  }
+
+  /**
+   * Converts a frequency expression into cronExpression that is usable by quartz.
+   * Supports frequency expressions with the following resolutions: minutes, hours, days.
+   * Example conversions:
+   * '10m' -> '*{@literal /}10 * * * ?'
+   * '3d' -> '0 0 *{@literal /}3 * ?'
+   *
+   * @return a cron expression
+   */
+  @VisibleForTesting
+  static String toCronExpr(String frequency) {
+    Preconditions.checkArgument(!Strings.isNullOrEmpty(frequency));
+    // remove all whitespace
+    frequency = frequency.replaceAll("\\s+", "");
+    Preconditions.checkArgument(frequency.length() >= 0);
+
+    frequency = frequency.toLowerCase();
+
+    String value = frequency.substring(0, frequency.length() - 1);
+    Preconditions.checkArgument(StringUtils.isNumeric(value));
+    Integer parsedValue = Integer.valueOf(value);
+    Preconditions.checkArgument(parsedValue > 0);
+
+    String everyN = String.format("*/%s", value);
```

Comment:
since we rely on the fact that it is in fact scheduled every N, we need to check that this divides evenly into the unit (for example, don't allow "13m" or "19h"), though maybe that happens elsewhere in the code.